core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java

```diff
@@ -61,10 +93,46 @@ public WorkerNetAddress getWorker(GetWorkerOptions options) {
       return null;
     }
     long randomLong = randomInCapacity(totalCapacity.get());
-    return rangeStartMap.floorEntry(randomLong).getValue().getNetAddress();
+    WorkerNetAddress targetWorker = rangeStartMap.floorEntry(randomLong).getValue().getNetAddress();
+    cacheWorker(options, targetWorker);
+    return targetWorker;
   }
 
   protected long randomInCapacity(long totalCapacity) {
     return ThreadLocalRandom.current().nextLong(totalCapacity);
   }
+
+  protected WorkerNetAddress findCacheWorker(GetWorkerOptions options) {
+    List<WorkerNetAddress> cacheCandidateList =
+        mBlockLocationCache.getIfPresent(options.getBlockInfo().getBlockId());
+    if (cacheCandidateList != null && mMaxReplicaSize > 0) {
+      Set<WorkerNetAddress> eligibleAddresses = new HashSet<>();
+      for (BlockWorkerInfo info : options.getBlockWorkerInfos()) {
+        eligibleAddresses.add(info.getNetAddress());
+      }
+      List<WorkerNetAddress> eligibleCacheList =
+          cacheCandidateList.stream().filter(eligibleAddresses::contains)
+              .collect(Collectors.toList());
+      if (eligibleCacheList.size() >= mMaxReplicaSize) {
+        int index = ThreadLocalRandom.current().nextInt(eligibleCacheList.size());
+        return eligibleCacheList.get(index);
+      }
+    }
+    return null;
+  }
+
+  protected void cacheWorker(GetWorkerOptions options, WorkerNetAddress targetWorker) {
```

Comment:
nit: I might prefer to make this function more straightforward, your call
```
   protect void addWorkerToCache(long blockId, WorderNetAddress workerAddress)
```

Reply:
@beinan @maobaolong done with your suggestion, please review again.